src/Middleware/RetryMiddleware.php

```diff
@@ -74,121 +76,91 @@ public function __construct(
     public function __invoke(Call $call, array $options)
     {
         $nextHandler = $this->nextHandler;
+        $retrySettings = $this->retrier->getRetrySettings();
 
         if (!isset($options['timeoutMillis'])) {
             // default to "noRetriesRpcTimeoutMillis" when retries are disabled, otherwise use "initialRpcTimeoutMillis"
-            if (!$this->retrySettings->retriesEnabled() && $this->retrySettings->getNoRetriesRpcTimeoutMillis() > 0) {
-                $options['timeoutMillis'] = $this->retrySettings->getNoRetriesRpcTimeoutMillis();
-            } elseif ($this->retrySettings->getInitialRpcTimeoutMillis() > 0) {
-                $options['timeoutMillis'] = $this->retrySettings->getInitialRpcTimeoutMillis();
+            if (!$retrySettings->retriesEnabled() && $retrySettings->getNoRetriesRpcTimeoutMillis() > 0) {
+                $options['timeoutMillis'] = $retrySettings->getNoRetriesRpcTimeoutMillis();
+            } elseif ($retrySettings->getInitialRpcTimeoutMillis() > 0) {
+                $options['timeoutMillis'] = $retrySettings->getInitialRpcTimeoutMillis();
             }
         }
 
         // Call the handler immediately if retry settings are disabled.
-        if (!$this->retrySettings->retriesEnabled()) {
+        if (!$retrySettings->retriesEnabled()) {
             return $nextHandler($call, $options);
         }
 
-        return $nextHandler($call, $options)->then(null, function ($e) use ($call, $options) {
-            $retryFunction = $this->getRetryFunction();
-
-            // If the number of retries has surpassed the max allowed retries
-            // then throw the exception as we normally would.
-            // If the maxRetries is set to 0, then we don't check this condition.
-            if (0 !== $this->retrySettings->getMaxRetries()
-                && $this->retryAttempts >= $this->retrySettings->getMaxRetries()
-            ) {
-                throw $e;
-            }
-            // If the retry function returns false then throw the
-            // exception as we normally would.
-            if (!$retryFunction($e, $options)) {
-                throw $e;
+        return $nextHandler($call, $options)->then(null, function ($exception) use ($call, $options) {
+            if (!$this->retrier->isRetryable($exception)) {
+                throw $exception;
             }
 
             // Retry function returned true, so we attempt another retry
-            return $this->retry($call, $options, $e->getStatus());
+            return $this->retry($call, $options, $exception);
         });
     }
 
     /**
      * @param Call $call
      * @param array $options
-     * @param string $status
+     * @param ApiException $exception
      *
      * @return PromiseInterface
      * @throws ApiException
      */
-    private function retry(Call $call, array $options, string $status)
+    private function retry(Call $call, array $options, ApiException $exception)
```

Comment:
Shouldn't this entire function live in the `Retrier`? That would also remove the need for so many public methods in the Retrier.

Reply:
@bshaffer This function uses `GuzzleHttp\Promise\PromiseInterface`. We wanted the `Retrier` to be as generic as possible. Thats why I did not move this functionality to the `Retrier` class.